infrastructure/src/main/java/org/corfudb/infrastructure/management/CompleteGraphAdvisor.java

```diff
@@ -0,0 +1,149 @@
+ * of the Corfu cluster is a fully connected network (i.e. complete graph) in which there is an
+ * active link amongst all nodes in the cluster. Failed and Healed nodes are recommended so that
+ * the cluster remains fully connected.
+ * <p>
+ * Created by Sam Behnam on 10/27/18.
+ */
+@Slf4j
+public class CompleteGraphAdvisor implements ClusterAdvisor {
+
+    private static final ClusterType CLUSTER_TYPE = ClusterType.COMPLETE_GRAPH;
+
+    private final String localEndpoint;
+
+    public CompleteGraphAdvisor(String localEndpoint) {
+        this.localEndpoint = localEndpoint;
+    }
+
+    @Override
+    public ClusterType getType() {
+        return CLUSTER_TYPE;
+    }
+
+    /**
+     * Provides list of servers from a given layout(epoch) that this implementation of
+     * COMPLETE_GRAPH algorithm has determined as failed. The implementation of the
+     * algorithm in this method is an approach by executing the following steps:
+     * <p>
+     * The failed node is the recommendation of this strategy which their removal
+     * from cluster will lead to a fully connected cluster.
+     * <p>
+     * To find a failed node:
+     * - get decision maker
+     * - find a failed node. Check if decision maker is not equal to failed node and failed node is not in unresponsive
+     * servers already.
+     *
+     * @param clusterState        represents the state of connectivity amongst the Corfu cluster
+     *                            nodes from a node's perspective.
+     * @param unresponsiveServers list of unresponsive servers.
+     * @return a server considered as failed according to the underlying strategy.
+     */
+    @Override
+    public Optional<NodeRank> failedServer(
+            ClusterState clusterState, List<String> unresponsiveServers) {
+
+        log.trace("Detecting failed nodes for: ClusterState= {} unresponsive servers= {}",
+                clusterState, unresponsiveServers
+        );
+
+        ClusterGraph graph = ClusterGraph.toClusterGraph(clusterState, localEndpoint);
+
+        ClusterGraph symmetric = graph.toSymmetric();
+        Optional<NodeRank> maybeDecisionMaker = symmetric.getDecisionMaker();
+
+        if (!maybeDecisionMaker.isPresent()) {
+            log.error("Decision maker not found for graph: {}", symmetric.toJson());
+            return Optional.empty();
+        }
+
+        NodeRank decisionMaker = maybeDecisionMaker.get();
+        if (!decisionMaker.is(localEndpoint)) {
+            String message = "The node can't be a decision maker, skip operation. Decision maker node is: {}";
+            log.debug(message, decisionMaker);
+            return Optional.empty();
+        }
+
+        Optional<NodeRank> maybeFailedNode = symmetric.findFailedNode();
+        if (!maybeFailedNode.isPresent()) {
+            return Optional.empty();
+        }
+
+        NodeRank failedNode = maybeFailedNode.get();
+
+        if (decisionMaker.equals(failedNode)) {
```

Comment:
What happens in this case?

Reply:
Just warning in the log, nothing happens. Not enough information to make a decision